File: client-android/serenada-core/src/main/java/app/serenada/core/call/AudioLevelMonitor.kt
package app.serenada.core.call

import kotlin.math.ln

/**
 * Stateful smoothing pipeline that turns a raw audio level (0..1, e.g. as
 * reported by WebRTC's `audioLevel` stat) into a perceptual 0..1 value
 * suitable for driving a voice-activity indicator.
 *
 * Mirrors the web SDK's `AudioLevelMonitor`:
 *  - Map RMS to dBFS, then linearly map [-60, -15] dBFS → [0, 1].
 *  - Apply asymmetric EMA: fast attack so bars react instantly to speech,
 *    slow release so they don't twitch off between syllables.
 *
 * Not thread-safe. Call [update] from a single thread (the same one that
 * reads [level]).
 */
internal class AudioLevelMonitor {
    private var smoothedLevel: Float = 0f

    val level: Float get() = smoothedLevel

    /** Submit a raw 0..1 level and return the smoothed value. */
    fun update(rawLevel: Float): Float {
        val raw = rawLevel.coerceIn(0f, 1f)
        val dbfs = if (raw > 0f) (20.0 * ln(raw.toDouble()) / LN_10).toFloat() else NOISE_FLOOR_DB
        val target = ((dbfs - NOISE_FLOOR_DB) / (SPEECH_PEAK_DB - NOISE_FLOOR_DB))
            .coerceIn(0f, 1f)
        val smoothing = if (target > smoothedLevel) ATTACK_SMOOTHING else RELEASE_SMOOTHING
        smoothedLevel = smoothedLevel * smoothing + target * (1f - smoothing)
        return smoothedLevel
    }

    fun reset() { smoothedLevel = 0f }

    companion object {
        /** Update interval that pairs with the indicator's CSS-equivalent transition. */
        const val UPDATE_INTERVAL_MS: Long = 100L
        /** dBFS at or below which the indicator reads zero. */
        const val NOISE_FLOOR_DB: Float = -60f
        /** dBFS at which the indicator reads full. Speech peaks around -20 to -15 dBFS. */
        const val SPEECH_PEAK_DB: Float = -15f
        /** Smoothing factor when level is rising. Lower = snappier attack. */
        const val ATTACK_SMOOTHING: Float = 0.4f
        /** Smoothing factor when level is falling. Higher = slower release. */
        const val RELEASE_SMOOTHING: Float = 0.7f
        private val LN_10 = ln(10.0)
    }
}

File: client-ios/SerenadaCore/Sources/Call/AudioLevelMonitor.swift
import Foundation

/// Stateful smoothing pipeline that turns a raw audio level (0..1, e.g. as
/// reported by WebRTC's `audioLevel` stat) into a perceptual 0..1 value
/// suitable for driving a voice-activity indicator.
///
/// Mirrors the web SDK's `AudioLevelMonitor` and the Android SDK's
/// `AudioLevelMonitor.kt`:
///  - Map RMS to dBFS, then linearly map [-60, -15] dBFS → [0, 1].
///  - Apply asymmetric EMA: fast attack so bars react instantly to speech,
///    slow release so they don't twitch off between syllables.
///
/// Not thread-safe. Use from a single thread (`@MainActor` in the SDK).
final class AudioLevelMonitor {
    /// Update interval that pairs with the indicator's animation duration.
    static let updateIntervalSeconds: Double = 0.1
    /// dBFS at or below which the indicator reads zero.
    static let noiseFloorDb: Float = -60
    /// dBFS at which the indicator reads full. Speech peaks around -20 to -15 dBFS.
    static let speechPeakDb: Float = -15
    /// Smoothing factor when level is rising. Lower = snappier attack.
    static let attackSmoothing: Float = 0.4
    /// Smoothing factor when level is falling. Higher = slower release.
    static let releaseSmoothing: Float = 0.7

    private(set) var level: Float = 0

    /// Submit a raw 0..1 level and return the smoothed value.
    @discardableResult
    func update(rawLevel: Float) -> Float {
        // `max`/`min` use `<` comparisons that propagate NaN unchanged, so a
        // non-finite input would slip through the clamp and pin the indicator
        // to a garbage value. Treat anything non-finite as silence.
        let sanitized = rawLevel.isFinite ? rawLevel : 0
        let raw = max(0, min(1, sanitized))
        let dbfs: Float = raw > 0 ? 20 * log10f(raw) : Self.noiseFloorDb
        let span = Self.speechPeakDb - Self.noiseFloorDb
        let target = max(0, min(1, (dbfs - Self.noiseFloorDb) / span))
        let smoothing = target > level ? Self.attackSmoothing : Self.releaseSmoothing
        level = level * smoothing + target * (1 - smoothing)
        return level
    }

    func reset() { level = 0 }
}
